### ISSUE TITLE
[META] ARM encoders, mobile/embedded devices

### ISSUE THREAD

**Issue by ehfd** [selkies-project/selkies#50]
Can be AI-driven: Yes, as long as all cases are tested carefully in real-life by people as well as automated tests. However, **expert guidance is required due to working with Linux intrinsics**.
Expected to be workable with: Fable 5, GPT-5.6 Sol, or equivalent (max thinking, requires expert guidance, especially with Opus 5, Kimi-K3, or under).

---

Note that we need to implement this on pixelflux and pcmflux, not GStreamer.

Encoders and screen-capturing interfaces exist from various embedded operating systems, including Raspberry Pi and NVIDIA Jetson. [gst-omx](https://gitlab.freedesktop.org/gstreamer/gstreamer/-/tree/main/subprojects/gst-omx) is an example of an encoder usable for Raspberry Pi and other ARM SoCs. `v4l2codecs` in gst-bad enables encoding support for Video4Linux as well.
While support for Android or iOS is probably out of our scope, ARM Linux support should be useful.

Relevant files: The equivalent of `omxh264enc` and/or `v4l2h264enc`.

**Comment by ehfd** [comment]
https://github.com/m1k1o/neko uses `v4l2h264enc` for Raspberry Pi.

**Comment by ehfd** [comment]
@ndufresne, maintainer of the v4l2 plugin of GStreamer Good Plugins:

```
yeah, so I'd say, stick with VP8/H264 (and HEVC perhaps)
I've seen some activity (of VP9 encoders) on Exynos side, after half a decade of silence, might be good news
as for stateless/rk (Rockchip), I expect one or two more years before that is ready
Media CODEC in Linux are still in catchup mode
```

VP9 encoders are out of the table for at least a year. Thus, I am focusing on H.264 and VP8, with H.265 to be back on the table if we land WebTransport.

https://gitlab.freedesktop.org/gstreamer/gst-plugins-good/-/commit/040e25218ac4d7cf50bcfd4651300b4bdfa629cb

**Comment by ehfd** [comment]
From https://github.com/selkies-project/docker-nvidia-egl-desktop/issues/23: @brianlmerritt

NVIDIA Jetson:

1. Compiling GStreamer for AARCH64 (addons/gstreamer)
2. Add nvv4l2h264enc/nvv4l2vp9enc and test
3. Verify all the pipeline works
Would be trivial for newbies.

https://docs.nvidia.com/jetson/archives/r35.2.1/DeveloperGuide/text/SD/Multimedia/AcceleratedGstreamer.html

Nicholas Dufresne (ndufresne) at #gstreamer IRC oftc.net should be experienced in this regard.

**Comment by ehfd** [comment]
GST-OMX is deprecated. Stick with v4l2.

**Comment by ehfd** [comment]
https://github.blog/changelog/2025-01-16-linux-arm64-hosted-runners-now-available-for-free-in-public-repositories-public-preview/

This will finally pave the way.

**Comment by ehfd** [comment]
Tegra is supported in https://github.com/selkies-project/pixelflux/pull/28.

**Comment by ndufresne** [comment]
Vulkan Video is likely the best successor to VA and V4L on Linux in the long term.

**Comment by ehfd** [comment]
> Vulkan Video is likely the best successor to VA and V4L on Linux in the long term.

@ndufresne Hi!

Yes, I agree, but Vulkan Video still has more room to mature, I think.

**Comment by ndufresne** [comment]
Right, though worth keeping track. It's the only acceleration supported by waypipe project fwiw.

**Comment by ehfd** [comment]
`v4l2` codec support in https://github.com/selkies-project/pixelflux/pull/35.

**Comment by ehfd** [comment]
We'll follow up with IBM POWER and Vulkan video codecs in another meta-issue. ARM encoders and embedded devices support were implemented.
